File: aoc23swift/aoc23swift/day11.swift
// Advent of Code 2023
// Dr Lee A. Christie
// @0x1ac@techhub.social

import Foundation

private func read_stars(example: Bool) throws -> (Int, Int, [(Int, Int)]) {
    let input = readInput(day: 11, example: example)
    var stars: [(Int, Int)] = []
    var width = 0
    var height = 0
    for (y, line) in input.split(separator: "\n").enumerated() {
        for (x, char) in line.enumerated() {
            if char == "#" {
                stars.append((x, y))
            }
            if x >= width {
                width = x + 1
            }
        }
        if y >= height {
            height = y + 1
        }
    }
    return (width, height, stars)
}

private func find_empty(width: Int, height: Int, stars: [(Int, Int)]) -> (Set<Int>, Set<Int>) {
    var empty_xs: Set<Int> = Set<Int>(0..<width)
    var empty_ys: Set<Int> = Set<Int>(0..<height)
    for (x, y) in stars {
        empty_xs.remove(x)
        empty_ys.remove(y)
    }
    return (empty_xs, empty_ys)
}

private func count_smaller(value: Int, others: Set<Int>) -> Int {
    var rv = 0
    for other in others {
        if other < value {
            rv += 1
        }
    }
    return rv
}

private func calculate_shift(dimension: Int, empty: Set<Int>) -> [Int:Int] {
    var rv: [Int:Int] = [:]
    for x in 0..<dimension {
        rv[x] = count_smaller(value: x, others: empty)
    }
    return rv
}

private func apply_shift(stars: [(Int, Int)], x_shift: [Int:Int], y_shift: [Int:Int], multiplier: Int) -> [(Int, Int)] {
    let added = multiplier - 1;
    var rv: [(Int, Int)] = []
    for (x, y) in stars {
        rv.append((x + added * x_shift[x]!, y + added * y_shift[y]!))
    }
    return rv
}

private func manhattan_distance(from: (Int, Int), to: (Int, Int)) -> Int {
    let (x0, y0) = from
    let (x1, y1) = to
    return abs(x0 - x1) + abs(y0 - y1)
}

private func calc_total_distance(stars: [(Int, Int)]) -> Int {
    var rv = 0
    for i in 0 ..< stars.count-1 {
        for j in i+1 ..< stars.count {
            rv += manhattan_distance(from: stars[i], to: stars[j])
        }
    }
    return rv
}

func solve11() throws {
    
    print("Advent of Code 2023")
    print("Day 11")
    
    // read the stars from the file
    let (width, height, stars) = try read_stars(example: false)
    
    // find the empty rows and columns
    let (empty_xs, empty_ys) = find_empty(width: width, height: height, stars: stars)
    
    // calculate how much to shift the stars by
    let x_shift = calculate_shift(dimension: width, empty: empty_xs)
    let y_shift = calculate_shift(dimension: height, empty: empty_ys)
    
    // shift the stars with multiplier x2 and calculate total distance
    let moved_stars_2 = apply_shift(stars: stars, x_shift: x_shift, y_shift: y_shift, multiplier: 2)
    var total_distance_2 = calc_total_distance(stars: moved_stars_2)
    
    print("Part 1: \(total_distance_2)")
    assert(9543156 == total_distance_2)
    
    // shift the stars with multiplier x1000000 and calculate total distance
    let moved_stars_1000000 = apply_shift(stars: stars, x_shift: x_shift, y_shift: y_shift, multiplier: 1000000)
    var total_distance_1000000 = calc_total_distance(stars: moved_stars_1000000)
        
    print("Part 2: \(total_distance_1000000)")
    assert(625243292686 == total_distance_1000000)
    
}

File: aoc23js/day11.js
// Advent of Code 2023
// Dr Lee A. Christie
// @0x1ac@techhub.social

const assert = require("assert");
const puzzle = require("./puzzle");

function read_stars(example) {

    const input = puzzle.inputLines(11, example);
    var width = 0;
    var height = 0;
    var y = 0;
    var stars = [];
    input.forEach((line) => {
        var x = 0;
        line.split("").forEach((char) => {
            if (char === "#") {
                stars.push([x, y]);
            }
            x++;
            if (x > width)
                width = x;
        });
        y++;
        if (y > height)
            height = y;
    });

    return [width, height, stars];

}

function find_empty(width, height, stars) {

    var empty_xs = new Set();
    for (var x = 0; x < width; x++)
        empty_xs.add(x);

    var empty_ys = new Set();
    for (var y = 0; y < width; y++)
        empty_ys.add(y);

    stars.forEach(([x, y]) => {
        empty_xs.delete(x);
        empty_ys.delete(y);
    });
    return [empty_xs, empty_ys];

}

function count_smaller(value, others) {
    var rv = 0;
    others.forEach((other) => {
        if (other < value)
            rv++;
    });
    return rv;
}

function calculate_shift(dimension, empty) {
    var rv = new Map();
    for (var i = 0; i < dimension; i++)
        rv[i] = count_smaller(i, empty);
    return rv;
}

function apply_shift(stars, x_shift, y_shift, multiplier) {
    let added = multiplier - 1;
    var rv = [];
    stars.forEach(([x, y]) => {
        rv.push([x + added * x_shift[x], y + added * y_shift[y]]);
    });
    return rv;
}

function manhattan_distance(from, to) {
    let [x0, y0] = from;
    let [x1, y1] = to;
    return Math.abs(x0 - x1) + Math.abs(y0 - y1);
}

function calc_total_distance(stars) {
    var rv = 0;
    for (var i = 0; i < stars.length - 1; i++) {
        for (var j = i + 1; j < stars.length; j++) {
            rv += manhattan_distance(stars[i], stars[j]);
        }
    }
    return rv;
}

function solve() {

    console.log("Advent of Code 2023");
    console.log('Day 11');

    // read the stars from the file
    let [width, height, stars] = read_stars(false);

    // find the empty rows and columns
    let [empty_xs, empty_ys] = find_empty(width, height, stars);

    // calculate how much to shift the stars by
    let x_shift = calculate_shift(width, empty_xs);
    let y_shift = calculate_shift(width, empty_ys);

    // shift the stars with multiplier x2 and calculate total distance
    let moved_stars_2 = apply_shift(stars, x_shift, y_shift, 2);
    let total_distance_2 = calc_total_distance(moved_stars_2);

    console.log(`Part 1: ${total_distance_2}`);
    assert(9543156 === total_distance_2);

    // shift the stars with multiplier x1000000 and calculate total distance
    let moved_stars_1000000 = apply_shift(stars, x_shift, y_shift, 1000000);
    let total_distance_1000000 = calc_total_distance(moved_stars_1000000);

    console.log(`Part 2: ${total_distance_1000000}`);
    assert(625243292686 === total_distance_1000000);

}

module.exports = solve;
